### PR TITLE
Refactor: use ctx_mailbox(Context)

### LINKED ISSUE #2762: Use more ctx_mailbox
In #2761 I introduced a new API, `struct Mailbox *ctx_mailbox(struct Context* ctx)`, which has the sole purpose to avoid boilerplate like `if (Context && Context->mailbox)`.

I would like to use the API more extensively, whenever `Context->mailbox` is used. This needs to be done with care as `Context` is a global variable and its contents might change when calling other functions.

This would be a nice small refactoring task.

### PR DESCRIPTION
**Pull request by rprimus**
* Refactor (Context && Context->mailbox) to ctx_mailbox(Context)
* Resolves #2762

### CODE REVIEW

**Comment by gahr** [comment]
This is great, thanks! A next step could be to change instaces of
* check `ctx_mailbox(Context)`
* conditionally use `Context->mailbox`
 
into instances of

* `(const) Mailbox * m = ctx_mailbox(Context);`
* conditionally use `m`

**Comment by rprimus** [comment]
Thu Dec 17 17:02:35 GMT 2020

There are 12 occurences of:
```
 if (!Context || !Context->mailbox)
```
eg `pattern/compile.c:177`

For consistency, should these be changed to:
```
! ctx_mailbox(Context)
```

**Comment by rprimus** [comment]
Fri Dec 18 11:26:49 GMT 2020

In index.c:694, where is (the global) Context initialised and how is Context->mailbox different from the *m that is passed to the function (change_folder_mailbox)?

**Comment by gahr** [comment]
`Context->mailbox` is the current mailbox, `change_folder_mailbox` gets passed the target mailbox you want to open.

**Comment by rprimus** [comment]
Sat Dec 19 04:05:50 GMT 2020

In the `Mailbox` struct definition (`core/mailbox.h`), all numeric values are `int`s.
```
  int msg_count;                      ///< Total number of messages
```

I've noticed (eg in index.c:788), there are comparisons between these struct member  `int` values and `size_t` values.
```
          for (size_t i = 0; i < m->msg_count; i++)
```

What's the view on these?

**Comment by rprimus** [comment]
Sat Dec 19 09:13:46 GMT 2020

Should be all done.  Have (mosty) left `index.c:mutt_index_menu` out of 2nd stage refactor (conditionally use `m`).

Locally, build is successful, however, not `make all-test` - link failure for `pattern/{pattern,compile}.c`: `_ctx_mailbox` symbol: "not found".

Looking into how `Makefile.sutosetup` should be updated.

**Comment by flatcap** [comment]
The fix I've just pushed is a bit ugly (no alternative).
We've got some tests for the Pattern code, but that code references `struct Context` leading to some unpleasant dependencies.

**Comment by rprimus** [comment]
Sat Dec 19 14:31:43 GMT 2020

@flatcap - Thank you for the build fix!

File: index.c [hunk 1]
```diff
@@ -689,23 +690,25 @@ static void change_folder_mailbox(struct Menu *menu, struct Mailbox *m, int *old
   if (!m)
     return;
 
+  struct Mailbox *cm = ctx_mailbox(Context);
```
Comment (flatcap) [inline index.c:693]: Please can you rename this to `m_ctx`.
I've given all the "main" structs standard names:
- `struct Mailbox *m`
- `struct Email *e`
etc

Where there are multiple instances, I've added suffixes, e.g. `m_sel`, `m_cur`.

File: monitor.c [hunk 2]
```diff
@@ -542,7 +543,9 @@ int mutt_monitor_remove(struct Mailbox *m)
     goto cleanup;
   }
 
-  if (Context && Context->mailbox)
+  struct Mailbox *cm = ctx_mailbox(Context);
```
Comment (flatcap) [inline monitor.c:546]: Please can you rename this to `m_ctx`

File: sidebar/window.c [hunk 3]
```diff
@@ -288,8 +289,9 @@ static const char *sidebar_format_str(char *buf, size_t buflen, size_t col, int
   if (!m)
     return src;
 
-  bool c = Context && Context->mailbox &&
-           mutt_str_equal(Context->mailbox->realpath, m->realpath);
+  struct Mailbox *cm = ctx_mailbox(Context);
```
Comment (flatcap) [inline sidebar/window.c:292]: Please can you rename this to `m_ctx`

File: test/pattern/dummy.c [hunk 4]
```diff
@@ -259,3 +262,8 @@ void mutt_buffer_select_file(struct Buffer *file, SelectFileFlags flags,
                              char ***files, int *numfiles)
 {
 }
+
+struct Mailbox *ctx_mailbox(struct Context *ctx)
+{
+  return Context ? Context->mailbox : NULL;
+}
```
Comment (flatcap) [inline test/pattern/dummy.c:269]: Please can you add a Doxygen function comment block (for complete-ness).
It doesn't need to say much.

**Review by flatcap — CHANGES_REQUESTED** [review]
Looks good at a first glance.
I'll have a deeper dive, later.

Thanks

File: sidebar/window.c (hunk 3 above)
Comment (rprimus) [reply sidebar/window.c]: Mon Dec 21 13:22:10 GMT 2020

Done.  See 15bd1a7

File: test/pattern/dummy.c (hunk 4 above)
Comment (rprimus) [reply test/pattern/dummy.c:276]: Mon Dec 21 13:37:12 GMT 2020

Done - see a12f1ff.

**Review by flatcap — APPROVED** [review]

**Pull request merged**